Implement Scratch Spaces (#1833)
JuliaLang/Pkg.jl: docs/src/spaces.md

# [**9.** Scratchspaces](@id Scratchspaces)

!!! compat "Julia 1.6"
    Pkg's Scratchspaces functionality requires at least Julia 1.6.

> `Pkg` can manage and lifecycle scratchspaces of temporary or readily-recreatable data.

Reviewer: It's not obvious to me what "lifecycle" is. Is that like garbage collect?

Author: Yes; lifecycle means how things get created and deleted.  Does this read better:

```
`Pkg` can manage and automatically garbage collect scratchspaces of temporary or readily-recreatable data.
```

These spaces can contain datasets, text, binaries, or any other kind of data that would be convenient to store, but which is non-fatal to have garbage collected if it has not been accessed recently, or if the owning package has been uninstalled.
As compared to [Artifacts](@ref), these containers of data are mutable and should be treated as ephemeral; all usage of scratchspaces should assume that the data stored within them could be gone by the next time your code is run.
In the current implementation, scratchspaces are removed during Pkg garbage collection if the space has not been accessed for a period of time (see the `spaces_cleanup_period` keyword argument to [Pkg.gc](@ref)), or if the owning package has been removed.
Users can also request a full wipe of all spaces to clean up unused disk space through `Pkg.Spaces.clear_spaces!()`.

> ## API overview

Reviewer: I wonder if this should also mention the case when a scrach file is " shared among multiple packages."

Reviewer: Or is that done by just leaving out the `pkg_uuid` arg in `get_space!`?

Author: Yes, precisely.  That's what is meant by a "global" space.

It provides a single interface for creating and getting previously-created spaces, either tied to a package by its UUID, or as a global scratch space that can be accessed by any package.
Here is an example where a package creates a space that is namespaced to its own UUID:

```julia
module SpaceExample
using Pkg, Pkg.Spaces

# This will be filled in inside `__init__()`
download_cache = ""

# Downloads a resource, stores it within a scratchspace
function download_dataset(url)
    fname = joinpath(download_cache, basename(url))
    if !isfile(fname)
        download(url, fname)
    end
    return fname
end

function __init__()
    global download_cache = @get_space!("downloaded_files")
end

end # module SpaceExample
```

Note that we initialize the `download_cache` within `__init__()` so that our packages are as relocatable as possible; we typically do not want to bake absolute paths into our precompiled files.
This makes use of the `@get_space!()` macro, which is identical to the `get_space!()` method, except it automatically determines the UUID of the calling module, if possible.
An equivalent (but more verbose) invocation is given here:
```julia
function __init__()

>     global download_cache = get_space!("downloaded_files"; pkg_uuid=Base.PkgId(@__MODULE__).uuid)

Reviewer: This feels quite a bit too internally to have in the docs. Couldn't `get_space!` take either a `Module` or a `UUID` and then dispatch accordingly?

Author: I'm confused; this is `get_space!()` taking in a `UUID` and dispatching accordingly?  We could of course have a method that takes in a `Module` as well.  Is that what you're suggesting?

Author: ah, you mean literally take either a `Module` or `UUID`, I understand now.  I mis-parsed your sentence.  :)

end
```

If a user wishes to manually delete a scratchspace, the method `delete_space!(key; pkg_uuid)` is the natural analog to `get_space!()`, however in general users will not need to do so, the spaces will be garbage collected by `Pkg` automatically.

For a full listing of docstrings and methods, see the [Spaces Reference](@ref) section.

> ## Usecases

Reviewer: ```suggestion
## Use cases
```

> Good usecases for a Pkg scratchspace include:

Reviewer: ```suggestion
Good use cases for a Pkg scratch space include:
```

* Caching downloads of files that must be routinely accessed and modified by a package.  Files that must be modified are a bad fit for the immutable [Artifacts](@ref) abstraction, and files can always be re-downloaded if the cache is wiped by the user.

* Generated data that depends on the characteristics of the host system.  Examples are compiled binaries, fontcache system font folder inspection output, generated CUDA bitcode files, etc...  Objects that would be difficult to compute off of the user's machine, and that can be recreated without user intervention are a great fit.

* Directories that should be shared between multiple packages in a single depot.  The space keying mechanism (explained above) makes it simple to provide scratch space that can be shared between different versions of a package, or even between different packages.  This allows packages to provide a scratch space where other packages can easily find the generated data, however the typical race condition warnings apply here; always design your access patterns assuming another process could be reading or writing to this scratch space at any time.

> Bad usecases for a Pkg scratchspace include (but are not limited to):

Reviewer: ```suggestion
Bad use cases for a Pkg scratch space include (but are not limited to):
```

## Frequently-Accessed Caching Questions

> Can I trigger data regeneration if the space is found to be empty/files are missing?

Yes, this is quite simple; just check the contents of the directory when you first call `get_space!()`, and if it's empty, run your generation function:

```julia
using Pkg, Pkg.Spaces

function get_dataset_dir()
    dataset_dir = @get_space!("dataset")
    if isempty(readdir(dataset_dir))
        perform_expensive_dataset_generation(dataset_dir)
    end
    return dataset_dir
end
```

> Can I create a scratchspace that is not shared across versions of my package?

Yes!  Make use of the `key` parameter and Pkg's ability to look up the current version of your package at compile-time:

```julia
module VersionSpecificExample
using Pkg, Pkg.Spaces

# Helpers to get current package UUID and VersionNumber
function get_uuid()

>     return Base.PkgId(@__MODULE__).uuid

Reviewer: Same with internal stuff here. I think we need an api that is not based on `PkgId`.

end
function get_version(uuid)

>     ctx = Pkg.Types.Context()

Reviewer: This is also internals of `Pkg`.

Author: If this is too internalsy, then we need to add an API to Pkg that gives you the version of the currently-running package.  I've needed to use that in a few packages now.

    uuid, entry = first(filter(((u, e),) -> u == uuid, ctx.env.manifest))
    return entry.version
end

# Get the current version at compile-time, that's fine it's not going to change. ;)

> const v = get_version(get_uuid())

Reviewer: If all the above was needed for this we should add an official API for it.

# This will be filled in by `__init__()`; it might change if we get deployed somewhere

> version_specific_space = ""

Reviewer: I wonder if it is in general better to advice people to use

```
version_specific_space = Ref{String}()
```

andb then 

```
version_specific_space[] = ...
```

in `__init__`?

That makes you able (for non isbitstypes) to actually check that the value has been set instead of not knowing whether you have the original `""` or it just got set to `""`.

Author: Hmmm, what situations are you thinking of that could allow for `version_specific_space` to _not_ be set?  It's within `__init__()`, so it should be guaranteed to either be set before any other (non-compile-time) code can run, or it will have thrown an error.

    # minor versions, but allows patch releases to share the same.
    global version_specific_space = @get_space!("data_for_version-$(v.major).$(v.minor)")
end

end # module
```

> Can I use a scratchspace as a temporary workspace, then turn it into an Artifact?

Yes!  Once you're satisfied with your dataset that has been cooking inside a space, and you're ready to share it with the world as an immutable artifact, you can use `create_artifact()` to create an artifact from the space, `archive_artifact()` to get a tarball that you can upload somewhere, and `bind_artifact!()` to write out an `Artifacts.toml` that allows others to download and use it:

```julia
using Pkg, Pkg.Spaces, Pkg.Artifacts
function export_space(space_name::String)
    space_dir = @get_space!(space_name)

    # Copy space directory over to an Artifact
    hash = create_artifact() do artifact_dir
        rm(artifact_dir)
        cp(space_dir, artifact_dir)
    end

    # Archive artifact out to a tarball
    mktempdir() do upload_dir
        tarball_path = joinpath(upload_dir, "$(space_name).tar.gz")
        tarball_hash = archive_artifact(hash, tarball_path)

>         # Upload tarball to a hosted site somewhere.  Note; this function does not exist:

Reviewer: Perhaps this should be written in a way that would make it easy to manually upload the tarball as well. Now with the `mktempdir` it would delete the directory when the scope exit so there is no point where you can manually do things.
